src/wp-includes/class-hierarchical-sort.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+/**
+ * Implements sorting posts by parent-child relationship.
+ *
+ * @package WordPress
+ * @since   6.8.0
+ */
+
+/**
+ * Sort post by hierarchy (parent-child relationship).
+ *
+ * @since 6.8.0
+ */
+class Hierarchical_Sort {
+
+	/**
+	 * Check if the request is eligible for hierarchical sorting.
+	 *
+	 * @param array $request The request data.
+	 *
+	 * @return bool Return true if the request is eligible for hierarchical sorting.
+	 */
+	public static function is_eligible( $request ) {
+		if ( ! isset( $request['orderby_hierarchy'] ) || true !== $request['orderby_hierarchy'] ) {
+			return false;
+		}
+
+		return true;
+	}
+
+	public static function run( $args ) {
+		$new_args = array_merge(
+			$args,
+			array(
+				'fields'         => 'id=>parent',
+				'posts_per_page' => -1,
+			)
+		);
```

Comment:
Modifying the query args with these params look similar to what is happening in the current WP Admin when hierarchical post types are queried without any order. See: https://github.com/WordPress/wordpress-develop/blob/9ad162e53ee9732c8ed63150650f7f9d3067a9d7/src/wp-admin/includes/post.php#L1294-L1300

I suspect we might need to update the `is_eligible()` method with the same check for an empty `orderby` query. I also wonder if we need to continue supporting the `menu_order` property as the default sorting for these types of queries?

For sites with a lot of posts, doing a query that returns all posts, e.g. `posts_per_page => -1` could be very slow, so there should be a way for this to be disabled.